pkg/validate/dynamic/platformworkloadidentityprofile.go

```diff
@@ -0,0 +1,130 @@
+package dynamic
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	sdkauthorization "github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v3"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/azuresdk/armauthorization"
+	"github.com/Azure/ARO-RP/pkg/util/rbac"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+func (dv *dynamic) ValidatePlatformWorkloadIdentityProfile(ctx context.Context, oc *api.OpenShiftCluster, platformWorkloadIdentityRolesByRoleName map[string]api.PlatformWorkloadIdentityRole, roleDefinitions armauthorization.RoleDefinitionsClient) error {
+	dv.log.Print("ValidatePlatformWorkloadIdentityProfile")
+
+	dv.platformIdentitiesActionsMap = map[string][]string{}
+	dv.platformIdentities = []api.PlatformWorkloadIdentity{}
+
+	for _, pwi := range oc.Properties.PlatformWorkloadIdentityProfile.PlatformWorkloadIdentities {
+		_, ok := platformWorkloadIdentityRolesByRoleName[pwi.OperatorName]
+		if ok {
+			dv.platformIdentitiesActionsMap[pwi.OperatorName] = nil
+			dv.platformIdentities = append(dv.platformIdentities, pwi)
+		}
+	}
+
+	// Check if any required platform identity is missing
+	if len(dv.platformIdentities) != len(platformWorkloadIdentityRolesByRoleName) {
+		requiredOperatorIdentities := []string{}
+		for _, role := range platformWorkloadIdentityRolesByRoleName {
+			requiredOperatorIdentities = append(requiredOperatorIdentities, role.OperatorName)
+		}
+		v := oc.Properties.ClusterProfile.Version
+		currentOpenShiftVersion, err := version.ParseVersion(oc.Properties.ClusterProfile.Version)
+		if err != nil {
+			return err
+		}
+		if oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo != nil {
+			upgradeableVersion, err := version.ParseVersion(string(*oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo))
+			if err != nil {
+				return err
+			}
+			if !upgradeableVersion.Eq(currentOpenShiftVersion) && !currentOpenShiftVersion.Lt(upgradeableVersion) {
+				v = fmt.Sprintf("%s or %s", v, string(*oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo))
+			}
+		}
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodePlatformWorkloadIdentityMismatch,
```

Comment:
So the failure here can be because of the platform identity validation failures. Those missing platform identities can be of either the current or upgradeableTo version.

Reply:
Okay, I see. So the assumption is that the cx should include `UpgradeableTo` and any new platform identities for that new OCP version in the same `PUT`/`PATCH` request?

Reply:
the cx should include UpgradeableTo only when all the new platform identities for that new OCP version are present in the cluster doc either via previous or the same PUT/PATCH request?